lib/Modem.ts

```diff
@@ -0,0 +1,829 @@
+		};
+
+		const serialPortOptions: SerialPortOptions = Object.assign(
+			// defaults
+			{
+				baudRate: 9600,
+				dataBits: 8,
+				stopBits: 1,
+				highWaterMark: 16384,
+				parity: 'none',
+				rtscts: false,
+				xon: false,
+				xoff: false
+			},
+			// options
+			options.serialPortOptions,
+			// overide options
+			{
+				path: targetDevice,
+				autoOpen: false
+			}
+		);
+
+		this.port = new SerialPort(serialPortOptions);
+		this.port.on('data', (data) => this._dataReceived(data));
+	}
+
+	/*
+	 * getter
+	 */
+
+	get targetDevice() {
+		return this.port.path;
+	}
+
+	get isOpen() {
+		return this.port.isOpen;
+	}
+
+	get mode() {
+		return this._mode;
+	}
+
+	get queueLength() {
+		return this.prioQueue.length + this.queue.length;
+	}
+
+	/*
+	 * command queue
+	 */
+
+	private pushToQue(cmd: Command | prTypes.CmdSequence, prio = false) {
+		if (prio) {
+			this.prioQueue.push(cmd);
+			return;
+		}
+
+		this.queue.push(cmd);
+	}
+
+	private async _execNext() {
+		if (this._isLocked || !this.port.isOpen) {
+			return;
+		}
+
+		const item = this.prioQueue.shift() || this.queue.shift();
+
+		if (!item) {
+			return;
+		}
+
+		this._isLocked = true;
+
+		if (item instanceof Command) {
+			await this._executeCMD(item);
+		} else {
+			for (const cmd of item.cmds) {
+				const result = await this._executeCMD(cmd);
+
+				if (result instanceof Error && item.cancelOnFailure) {
+					if (item.onFailed) {
+						item.onFailed(result);
+					}
+
+					break;
+				}
+			}
+		}
+
+		this._isLocked = false;
+		return item;
+	}
+
+	private async _executeCMD(cmd: Command) {
+		const result = await new Promise((resolve: (result: CommandResponse | Error) => void) => {
+			if (cmd.deprecated) {
+				return resolve(new Error('Command marked as deprecated'));
+			}
+
+			const write = `${cmd.ATCommand}\r`;
+
+			this.events.emitWriteToModem(write);
+			this.port.write(write);
+
+			const timeout = setTimeout(() => resolve(new Error('Command is timeouted!')), cmd.timeout);
+
+			this.events.once('onCommandResponse', (data) => {
+				clearTimeout(timeout);
+				resolve(data);
+			});
+		});
+
+		try {
+			cmd.callback(result);
+		} finally {
+			// not used
+		}
+
+		return result;
+	}
+
+	/*
+	 * handling for received data from the modem
+	 */
+
+	private _dataReceived(received: Buffer) {
+		this._receivedData += received.toString();
+		const parts = this._receivedData.split('\r\n');
+		this._receivedData = parts.pop() || '';
+
+		console.log(this._receivedData);
+
+		for (const part of parts) {
+			if (part.trim() !== '') {
+				this.events.emitDataReceived(part);
+			}
+
+			for (const newpart of part.split(/\r\n|\n|\r/)) {
+				if (newpart.trim() === '') {
+					continue;
+				}
+
+				// skip echoes from commands
+				if (newpart.startsWith('AT')) {
+					continue;
+				}
+
+				// new incomming SMS
+				if (newpart.startsWith('+CMTI:')) {
+					const smsID: string | undefined = newpart.substring(7).split(',')[1];
```

Comment:
why not do Number convert already there? then it is done only once

Reply:
Is hopefully improved in the next update